zkevm-circuits/src/evm_circuit/witness.rs

```diff
@@ -0,0 +1,522 @@
+    };
+    result
+}
+
+fn tx_convert(
+    randomness: Fp,
+    bytecode: &Bytecode,
+    tx: &bus_mapping::circuit_input_builder::Transaction,
+    ops_len: (usize, usize, usize),
+) -> Transaction<Fp> {
+    let mut result: Transaction<Fp> = Transaction::<Fp> {
+        calls: vec![Call {
+            id: 1,
+            is_root: true,
+            is_create: tx.is_create(),
+            opcode_source: RandomLinearCombination::random_linear_combine(
+                bytecode.hash.to_le_bytes(),
+                randomness,
+            ),
+        }],
+        ..Default::default()
+    };
+    result.calls = vec![Call {
+        id: 1,
+        is_root: true,
+        is_create: tx.is_create(),
+        opcode_source: RandomLinearCombination::random_linear_combine(
+            bytecode.hash.to_le_bytes(),
+            randomness,
+        ),
+    }];
+    for idx in 0..tx.steps().len() {
+        let cur_step = &tx.steps()[idx];
+        let prev_step = if idx == 0 {
+            None
+        } else {
+            Some(&tx.steps()[idx - 1])
+        };
+        result
+            .steps
+            .push(step_convert(prev_step, cur_step, ops_len));
+    }
+    result
+}
+
+pub fn block_convert(
+    bytecode: &bus_mapping::bytecode::Bytecode,
+    b: &bus_mapping::circuit_input_builder::Block,
+) -> Block<Fp> {
+    let randomness = Fp::rand();
+    let bytecode = bytecode.into();
+
+    // here stack_ops/memory_ops/etc are merged into a single array
+    // in EVM circuit, we need gc-sorted ops
+    let mut stack_ops = b.container.sorted_stack();
+    stack_ops.sort_by_key(|s| usize::from(s.gc()));
+    let mut memory_ops = b.container.sorted_memory();
+    memory_ops.sort_by_key(|s| usize::from(s.gc()));
+    let mut storage_ops = b.container.sorted_storage();
+    storage_ops.sort_by_key(|s| usize::from(s.gc()));
+
+    let mut block = Block {
+        randomness,
+        txs: b
+            .txs()
+            .iter()
+            .map(|tx| {
+                tx_convert(
+                    randomness,
+                    &bytecode,
+                    tx,
+                    (stack_ops.len(), memory_ops.len(), storage_ops.len()),
+                )
+            })
+            .collect(),
+        bytecodes: vec![bytecode],
+        ..Default::default()
+    };
+
+    block.rws.extend(stack_ops.iter().map(|s| Rw::Stack {
+        rw_counter: s.gc().into(),
+        is_write: s.op().rw().is_write(),
+        call_id: 1,
+        stack_pointer: usize::from(*s.op().address()),
+        value: *s.op().value(),
+    }));
+    block.rws.extend(memory_ops.iter().map(|s| Rw::Memory {
+        rw_counter: s.gc().into(),
+        is_write: s.op().rw().is_write(),
+        call_id: 1,
+        memory_address: u64::from_le_bytes(
+            s.op().address().to_le_bytes()[..8].try_into().unwrap(),
+        ),
+        byte: s.op().value(),
+    }));
+    // TODO add storage ops
+
+    block
+}
+
+#[allow(clippy::needless_borrow)]
```

Comment:
Is this trying to emphasize the borrow of `bytecode`? Maybe we could just remove the `&` at line 511 and 521?

Reply:
https://github.com/appliedzkp/zkevm-circuits/pull/239/commits/afd789503dfc0f3922ed14c1cd0c49004493ee40

Reply:
> Is this trying to emphasize the borrow of `bytecode`? Maybe we could just remove the `&` at line 511 and 521?

https://github.com/appliedzkp/zkevm-circuits/pull/239/commits/afd789503dfc0f3922ed14c1cd0c49004493ee40